pulsar/src/main/java/org/openconnectors/PulsarSource.java

```diff
@@ -0,0 +1,88 @@
+package org.openconnectors;
+
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.openconnectors.config.Config;
+import org.openconnectors.connect.ConnectorContext;
+import org.openconnectors.connect.PushSourceConnector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.function.Consumer;
+
+import static org.openconnectors.config.ConfigUtils.verifyExists;
+
+public class PulsarSource implements PushSourceConnector<Message> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PulsarSource.class);
+
+    private java.util.function.Consumer<Collection<Message>> consumeFunction;
+    PulsarClient client;
+    private org.apache.pulsar.client.api.Consumer consumer;
+
+    @Override
+    public void setConsumer(Consumer<Collection<Message>> consumeFunction) {
+        this.consumeFunction = consumeFunction;
+    }
+
+    @Override
+    public void initialize(ConnectorContext ctx) {
+
+    }
+
+    @Override
+    public void open(Config config) throws Exception {
+
+        verifyExists(config, PulsarConfigKeys.PULSAR_SOURCE_TOPIC);
+        verifyExists(config, PulsarConfigKeys.PULSAR_SOURCE_BROKER_ROOT_URL);
+        verifyExists(config, PulsarConfigKeys.PULSAR_SOURCE_SUBSCRIPTION);
+
+        String pulsarBrokerRootUrl = config.getString(PulsarConfigKeys.PULSAR_SOURCE_BROKER_ROOT_URL);
+        client = PulsarClient.create(pulsarBrokerRootUrl);
+
+        String topic = config.getString(PulsarConfigKeys.PULSAR_SOURCE_TOPIC);
+        String subscription = config.getString(PulsarConfigKeys.PULSAR_SOURCE_SUBSCRIPTION);
+        consumer = client.subscribe(topic, subscription);
+
+        start();
+    }
+
+    private void start() {
+        Thread runnerThread = new Thread(() -> {
+            try {
+
+                while (true) {
+                    // Wait for a message
+                    Message msg = consumer.receive();
+
+                    consumeFunction.accept(Collections.singleton(msg));
+
+                    // Acknowledge the message so that it can be deleted by broker
+                    consumer.acknowledge(msg);
```

Comment:
We should use `acknowledgeAsync(msg)` since there's no point in waiting for the future: the only reason for the ack to fail is if we're not currently connected. in that case, the message is going to replayed anyway.

Reply:
will change